common/cpp/src/google_smart_card_common/value_conversion.cc

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <google_smart_card_common/value_conversion.h>
+
+#include <stdint.h>
+
+#include <limits>
+#include <string>
+#include <utility>
+
+#include <google_smart_card_common/formatting.h>
+#include <google_smart_card_common/logging/logging.h>
+#include <google_smart_card_common/numeric_conversions.h>
+#include <google_smart_card_common/value.h>
+#include <google_smart_card_common/value_debug_dumping.h>
+
+namespace google_smart_card {
+
+namespace {
+
+constexpr char kErrorWrongType[] = "Expected value of type %s, instead got: %s";
+
+template <typename T>
+bool ConvertIntegerFromValue(Value value, const char* type_name, T* number,
+                             std::string* error_message) {
+  int64_t int64_number;
+  if (value.is_integer()) {
+    int64_number = value.GetInteger();
+  } else if (value.is_float()) {
+    // Non-32-bit numbers might arrive as floating-point `Value` objects, so
+    // attempt a cast in case the value lies within the range of precisely
+    // representable integers.
+    if (!CastDoubleToInt64(value.GetFloat(), &int64_number, error_message))
+      return false;
+  } else {
+    if (error_message) {
+      *error_message =
+          FormatPrintfTemplate(kErrorWrongType, Value::kIntegerTypeTitle,
+                               DebugDumpValueSanitized(value).c_str());
+    }
+    return false;
+  }
+  return CastInt64ToInteger(int64_number, type_name, number, error_message);
+}
+
+}  // namespace
+
+Value ConvertToValue(unsigned number) {
+  // The Standard doesn't give an upper bound on the `unsigned` size, so be on
+  // the safe side (despite that a typical implementation has it <=4 bytes).
+  GOOGLE_SMART_CARD_CHECK(
```

Comment:
Actually, folks, WDYT - given these error scenarios, should I also make the `ConvertToValue()` functions be able to report the error rather than crash?

This would also be consistent with the `ConvertFromValue()` interface.

Reply:
I think this is a good idea.

Reply:
Done